python/plugins/processing/algs/gdal/rasterize_over.py

```diff
@@ -23,36 +23,68 @@
 
 import os
 
-from qgis.core import QgsProcessingUtils
-
 from qgis.PyQt.QtGui import QIcon
 
+from qgis.core import (QgsRasterFileWriter,
+                       QgsProcessingParameterDefinition,
+                       QgsProcessingParameterFeatureSource,
+                       QgsProcessingParameterField,
+                       QgsProcessingParameterRasterLayer,
+                       QgsProcessingParameterNumber,
+                       QgsProcessingParameterString,
+                       QgsProcessingParameterBoolean)
 from processing.algs.gdal.GdalAlgorithm import GdalAlgorithm
 from processing.algs.gdal.GdalUtils import GdalUtils
 
-from processing.tools import dataobjects
-
 pluginPath = os.path.split(os.path.split(os.path.dirname(__file__))[0])[0]
 
 
 class rasterize_over(GdalAlgorithm):
 
     INPUT = 'INPUT'
-    INPUT_RASTER = 'INPUT_RASTER'
     FIELD = 'FIELD'
-
-    def icon(self):
-        return QIcon(os.path.join(pluginPath, 'images', 'gdaltools', 'rasterize.png'))
+    INPUT_RASTER = 'INPUT_RASTER'
+    ADD = 'ADD'
+    EXTRA = 'EXTRA'
+    BURN = 'BURN'
 
     def __init__(self):
         super().__init__()
 
     def initAlgorithm(self, config=None):
-        self.addParameter(ParameterVector(self.INPUT, self.tr('Input layer')))
-        self.addParameter(ParameterTableField(self.FIELD,
-                                              self.tr('Attribute field'), self.INPUT))
-        self.addParameter(ParameterRaster(self.INPUT_RASTER,
-                                          self.tr('Existing raster layer'), False))
+        self.units = [self.tr("Pixels"),
+                      self.tr("Georeferenced units")]
+
+        self.addParameter(QgsProcessingParameterFeatureSource(self.INPUT,
+                                                              self.tr('Input vector layer')))
+
+        self.addParameter(QgsProcessingParameterRasterLayer(self.INPUT_RASTER, self.tr('Input raster layer')))
+
+        self.addParameter(QgsProcessingParameterField(self.FIELD,
+                                                      self.tr('Field to use for burn in value'),
+                                                      None,
+                                                      self.INPUT,
+                                                      QgsProcessingParameterField.Numeric,
+                                                      optional=True))
+
+        self.addParameter(QgsProcessingParameterNumber(self.BURN,
```

Comment:
Wouldn't it be better to split this to two algorithms, to avoid the same issue that #32097 is addressing?

Reply:
@nyalldawson in my opinion the situation is similar but not exactly the same. The UI for that SAGA tool was too complicated (I think that this is not arguable) and moreover to make the tool work was necessary to enter a fake parameter (an empty reclassification table). Both things contributed to make the tool not very user friendly. If I'm not wrong this happens because for Processing tools based on description files (and using just description files) we cannot define any type of logic (i.e. If this parameter is set then ignore the other). On the other hand for GDAL based tools we can easily manage logic with parameters, in the case for gdal rasterize (and I guess for several other similar cases) the values to be transferred to the pixels can come from a column of the attribute table, or a fixed, user-defined value. The latter is used by default unless the former has been defined. This way of functioning is not ideal but is perfectly acceptable for a user, and so we are not obligated to split the tool. Anyway this is just my opinion, if we choose to split... I will split.

Reply:
Ok, let's get a second opinion from @alexbruy?

Reply:
Personally I preffer to have simpler algorithms with parameters specific for single task. Of course, we can add conditional logic to handle cases like "if paramA is present use it, otherwise use paramB" but IMHO this is error prone. Also algorithm UI becomes much complex, for example look at the import into PostGIS and some other algorithms, sometimes their dialogs even does not fit into screen and have a lot of parameters with not-so-obvious behavior.

I also think that is algorithm use different parameters for the same value, better to split it into two or use dynamic parameters if possible.

Can we use a data-defined parameter here?

Reply:
> Can we use a data-defined parameter here?

Nope - that's a (killer) feature solely available to the native algorithms.